omp/multigrid/amgx_pgm_kernels.cpp

```diff
@@ -62,22 +68,62 @@ namespace amgx_pgm {
 template <typename IndexType>
 void match_edge(std::shared_ptr<const OmpExecutor> exec,
                 const Array<IndexType> &strongest_neighbor,
-                Array<IndexType> &agg) GKO_NOT_IMPLEMENTED;
+                Array<IndexType> &agg)
+{
+    auto agg_vals = agg.get_data();
+    auto strongest_neighbor_vals = strongest_neighbor.get_const_data();
+#pragma omp parallel for
+    for (size_type i = 0; i < agg.get_num_elems(); i++) {
+        if (agg_vals[i] == -1) {
+            size_type neighbor = strongest_neighbor_vals[i];
+            if (neighbor != -1 && strongest_neighbor_vals[neighbor] == i) {
+                // Use the smaller index as agg point
+                auto group = min(i, neighbor);
+                agg_vals[i] = group;
+                agg_vals[neighbor] = group;
+            }
+        }
+    }
+}
 
 GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(GKO_DECLARE_AMGX_PGM_MATCH_EDGE_KERNEL);
 
 
 template <typename IndexType>
 void count_unagg(std::shared_ptr<const OmpExecutor> exec,
-                 const Array<IndexType> &agg,
-                 IndexType *num_unagg) GKO_NOT_IMPLEMENTED;
+                 const Array<IndexType> &agg, IndexType *num_unagg)
+{
+    IndexType unagg = 0;
+#pragma omp parallel for reduction(+ : unagg)
+    for (size_type i = 0; i < agg.get_num_elems(); i++) {
+        unagg += (agg.get_const_data()[i] == -1);
+    }
+    *num_unagg = unagg;
+}
 
 GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(GKO_DECLARE_AMGX_PGM_COUNT_UNAGG_KERNEL);
 
 
 template <typename IndexType>
 void renumber(std::shared_ptr<const OmpExecutor> exec, Array<IndexType> &agg,
-              IndexType *num_agg) GKO_NOT_IMPLEMENTED;
+              IndexType *num_agg)
+{
+    const auto num = agg.get_num_elems();
+    Array<IndexType> agg_map(exec, num + 1);
+    auto agg_vals = agg.get_data();
+    auto agg_map_vals = agg_map.get_data();
+    components::fill_array(exec, agg_map_vals, num + 1, zero<IndexType>());
```

Comment:
```suggestion
    agg_map_vals.fill(zero<IndexType>());
```

Reply:
I do not notice we add it in array header.
However, is it acceptable? Doesn't it lead the circular dependence?
I think we can use array in device folder because Array is header-only class.

Reply:
Good point, this will not work and create a circular dependency issue since `Array::fill` is implemented in the `core` module and not in the header.

Reply:
Yes, you are right. Forget about these fill changes that I asked for.

Reply:
Should I remove the fill member function out of array in this pr?